sys/chunked_bucket.js

```diff
@@ -0,0 +1,433 @@
+
+function range(N) {
+    return Array.apply(null, { length: N }).map(Number.call, Number);
+}
+
+function ChunkedBucket(options) {
+    this.log = options.log || function() {};
+}
+
+ChunkedBucket.prototype._sliceData = function(data) {
+    var slicer;
+    if (Buffer.isBuffer(data)) {
+        slicer = Buffer.prototype.slice;
+    } else {
+        data = data.toString();
+        slicer = String.prototype.substring;
+    }
+    var result = [];
+    for (var index = 0; index < data.length; index += CHUNK_SIZE) {
+        result.push(slicer.call(data, index, index + CHUNK_SIZE));
+    }
+    return result;
+};
+
+ChunkedBucket.prototype._joinData = function(dataParts) {
+    if (!dataParts || !dataParts.length) {
+        return '';
+    } else if (Buffer.isBuffer(dataParts[0])) {
+        return Buffer.concat(dataParts);
+    } else {
+        return String.prototype.concat.apply(dataParts[0], dataParts.slice(1));
+    }
+};
+
+ChunkedBucket.prototype._metaTableName = function(req) {
+    return req.params.bucket + '_meta';
+};
+
+ChunkedBucket.prototype._chunksTableName = function(req) {
+    return req.params.bucket + '_chunks';
+};
+
+ChunkedBucket.prototype._makeMetaSchema = function(opts) {
+    var schemaVersionMajor = 1;
+
+    return {
+        version: schemaVersionMajor * 1000 + (opts.version || 0),
+        options: {
+            compression: opts.compression || [
+                {
+                    algorithm: 'deflate',
+                    block_size: 1024
+                }
+            ]
+        },
+        attributes: {
+            key: opts.keyType || 'string',
+            rev: 'int',
+            tid: 'timeuuid',
+            num: 'int',
+            headers: 'json'
+        },
+        index: [
+            { attribute: 'key', type: 'hash' },
+            { attribute: 'rev', type: 'range', order: 'desc' },
+            { attribute: 'tid', type: 'range', order: 'desc' }
+        ]
+    };
+};
+
+ChunkedBucket.prototype._makeChunksSchema = function(opts) {
+    var schemaVersionMajor = 1;
+
+    return {
+        version: schemaVersionMajor * 1000 + (opts.version || 0),
+        options: {
+            compression: opts.compression || [
+                {
+                    algorithm: 'deflate',
+                    block_size: 1024
+                }
+            ]
+        },
+        attributes: {
+            key: opts.keyType || 'string',
+            chunk_id: 'int',
+            tid: 'timeuuid',
+            value: opts.valueType || 'blob'
+        },
+        index: [
+            { attribute: 'key', type: 'hash' },
+            { attribute: 'chunk_id', type: 'hash' },
+            { attribute: 'tid', type: 'range', order: 'desc' }
+        ]
+    };
+};
+
+ChunkedBucket.prototype.createBucket = function(restbase, req) {
+    var metaSchema = this._makeMetaSchema(req.body || {});
+    var dataSchema = this._makeChunksSchema(req.body || {});
+    metaSchema.table = this._metaTableName(req);
```

Comment:
We discussed using an extension attribute ("custom"?) in the table schema to record this information. That information can then be retrieved from the internal meta table using `GET /table/{name}`, which calls [getTableSchema](https://github.com/wikimedia/restbase-mod-table-cassandra/blob/master/index.js#L145). The advantage of doing it this ways is two less tables & one less keyspace per chunked bucket.

The disadvantage of this approach is the need to change the table storage backends to accept / ignore the custom attribute. But, perhaps we could just hard-code parameters for now, and only start storing them once we actually have a need to support bucket meta options?